### PR TITLE
Add signed num::NonZeroI* types

### PR DESCRIPTION
**Pull request by SimonSapin**
Multiple people have asked for them in https://github.com/rust-lang/rust/issues/49137. Given that the unsigned ones already exist, they are very easy to add and not an additional maintenance burden.

### CODE REVIEW

**Comment by SimonSapin** [comment]
The new types are stable in this PR, so let’s get team sign-off:

@rfcbot fcp merge

They don’t really need to be. But they are very similar to the existing unsigned non-zero integer types, so it seems to me that having them as unstable for a period of time is very unlikely to bring new information. Also, making the signed and unsigned type behave differently (beyond signedness) or have different API seems like it would be unfortunate, and the latter are already stable.

**Comment by Amanieu** [comment]
I'm in favor of this since it would make `NonZero*` consistent with `Atomic*`, which has both signed and unsigned variants even though you could get away with just one.

**Comment by xfix** [comment]
I think it would make sense to have some tests for `NonZeroI*` types.

**Review by sugar700 — CHANGES_REQUESTED** [review]
I think it would make sense to have some tests for NonZeroI* types. Check whether they can store negative values without causing undefined behaviour, and so on.

**Comment by Xaeroxe** [comment]
Thank you for this! [This is the happiest I've ever been to deprecate my own code.](https://github.com/Xaeroxe/nonzero_signed/pull/1)

**Comment by SimonSapin** [comment]
I’d appreciate some help. Maybe @alexcrichton?

I’m getting errors like

```rust
---- num/mod.rs - num::NonZeroI16 (line 33) stdout ----
error[E0412]: cannot find type `NonZeroI16` in module `std::num`
 --> num/mod.rs:35:39
  |
5 | assert_eq!(size_of::<Option<std::num::NonZeroI16>>(), size_of::<i16>());
  |                                       ^^^^^^^^^^ did you mean `NonZeroU16`?

thread 'num/mod.rs - num::NonZeroI16 (line 33)' panicked at 'couldn't compile the test', src/librustdoc/test.rs:321:13
```

It looks like a newly-added type cannot be found *in the doctests of that very type*.

Later in the output:

```
command did not execute successfully: "/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "8" "--release" "--features" "panic-unwind backtrace" "--manifest-path" "/home/simon/rust3/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
expected success, got: exit code: 101
```

… which includes `stage0`. The error would make some sense if the bootstrap compiler’s libcore is used to run the new libcore’s doctests, but why would that be the case when running simply `./x.py test`?

**Comment by SimonSapin** [comment]
Also, `./x.py test --test-args try-block/try-block-bad-type.rs` succeeds on my machine, even after a rebase on `master`, while the same test fails on Travis :/

**Comment by alexcrichton** [comment]
@SimonSapin hm something looks seriously wrong here! I wonder if this is a preexisting compiler bug you're running into. Can you reproduce the travis failure with `./src/ci/docker/run.sh x86_64-gnu-llvm-6.0`?

**Review by sugar700 — APPROVED** [review]

**Comment by SimonSapin** [comment]
@alexcrichton Yes, that command on my machine gives the same output as on Travis.

**Comment by SimonSapin** [comment]
eIf I tweak `src/test/ui/try-block/try-block-bad-type.stderr` to get past that error, I can get to the same error as https://github.com/rust-lang/rust/pull/57475#issuecomment-453291779 in Docker (and presumably on Travis). But then it is `./x.py test` on by Ubuntu 18.04 environment that fails `try-block-bad-type`.

**Comment by SimonSapin** [comment]
@oli-obk (CC https://github.com/rust-lang/rust/pull/54032), is the valid range for the purpose of `#[rustc_layout_scalar_valid_range_start(1)]` based on the "unsigned view" of some memory bits, or on the actual semantics of the type? Specifically, does it only exclude zero when used on a wrapper of a signed integer type like `i32`, or also negative values?

**Comment by oli-obk** [comment]
It works on the unsigned view. It uses the same representation as our compiler internal layout definition.

**Comment by SimonSapin** [comment]
> It looks like a newly-added type cannot be found in the doctests of that very type.

When running `./x.py test -vv`, I see (newlines added):

```
rustdoc command:
"/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc"
"--test"
"/home/simon/rust3/src/libcore/lib.rs"
"--crate-name"
"core"
"-L"
"dependency=/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps"
"-L"
"dependency=/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage1-std/release/deps"
"--test-args"
"NonZero"
"--test-args"
"--quiet"
"--extern"
"core=/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcore-852c25e755af0d0c.rlib"
"--extern"
"rand=/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librand-5423a3ce52cd7271.rlib"
"--cfg"
"stage1"
"--cfg"
"dox"
"--sysroot"
"/home/simon/rust3/build/x86_64-unknown-linux-gnu/stage1"
"-Z"
"force-unstable-if-unmarked"
"-Z"
"unstable-options"
"--crate-version"
"1.33.0-dev"
```

Note the presence of `--extern core=…`, but nothing similar for `std`. Changing `std::num` to `core::num` in the doctest seems to work around this issue.

Is it possible that doctests are using libcore compiled from the repository, but libstd from the bootstrap / stage0 compiler?

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -5,14 +5,14 @@
 use convert::TryFrom;
 use fmt;
 use intrinsics;
-use mem;
+use mem;9e448348b9
```
Comment (oli-obk) [inline src/libcore/num/mod.rs]: accidental paste

Comment (SimonSapin) [reply src/libcore/num/mod.rs]: Good catch, thanks.

**Comment by Centril** [comment]
Failed in rollup, https://github.com/rust-lang/rust/pull/57727#issuecomment-455517221.

@bors r-

**Comment by SimonSapin** [comment]
The set of impls printed (or at least their order) in the `error[E0277]` diagnostic varies between environments. The diff below would (likely) fix the `mingw-check` CI job like in https://github.com/rust-lang/rust/pull/57727#issuecomment-455517221, but break `x86_64-gnu-llvm-6.0` like in https://github.com/rust-lang/rust/pull/57475#issuecomment-453183765.

Any idea what’s going on?

```diff
diff --git a/src/test/ui/try-block/try-block-bad-type.stderr b/src/test/ui/try-block/try-block-bad-type.stderr
index c2f9e9b52b..4e964f4b83 100644
--- a/src/test/ui/try-block/try-block-bad-type.stderr
+++ b/src/test/ui/try-block/try-block-bad-type.stderr
@@ -5,9 +5,9 @@ LL |         Err("")?; //~ ERROR the trait bound `i32: std::convert::From<&str>`
    |         ^^^^^^^^ the trait `std::convert::From<&str>` is not implemented for `i32`
    |
    = help: the following implementations were found:
-             <i32 as std::convert::From<core::num::NonZeroI32>>
+             <i32 as std::convert::From<bool>>
              <i32 as std::convert::From<i16>>
-             <i32 as std::convert::From<i8>>
+             <i32 as std::convert::From<u16>>
              <i32 as std::convert::From<u8>>
            and 2 others
    = note: required by `std::convert::From::from`

**Comment by SimonSapin** [comment]
I pushed a commit that I hope will fix this by removing some non-determinism.

@estebank, could you have a look? Can `impl_candidates` in `report_similar_impl_candidates` be very large? With this commit we’re doing some work (`normalize` and `sort`) on all of it, not just the first 4 or 5 impls.

**Comment by estebank** [comment]
@SimonSapin it is _potentially_ huge, but in practice I would only imagine this being problematic in the case of generated code. But there are two things: 1) this is only relevant for failing code, so the cost only needs to be "human-speed", we're unlikely to hit problems there in any realistic scenario and 2) we might get away with changing `fn all_impls(self, def_id: DefId) -> Vec<DefId>` to return an `Iterator`. That wouldn't be a huge code change, nor performance change, but we're doing `iter` -> `collect` -> `iter` -> `filter_map` -> `collect` a few times on the same data :-|

I would go ahead with it, maybe adding a debug statement before and after to be able to check after the fact with a pathological case what the time is.

**Comment by xfix** [comment]
Sorting can be optimized by using heap sort instead of whatever sorting algorithm uses, but considering this is an error case, it's likely not worth doing.

**Pull request merged**